강원대 FE 유다연 - 1단계 로그인 기능 구현하기

daaoooy opened the pull request:
### [미션 02 : 1단계 - 로그인 기능 구현하기]

- [x]  본인만의 기준으로 일관된 코드를 작성하기
- [x]  기능 단위로 나누어 커밋하기

#### 로그인 - ID
- [x] ID 는 반드시 입력되어야 하고 이메일 형식이어야 함 
- [x] ID input field를 벗어났을 때, 이 조건이 충족되지 않으면 적절한 에러 메세지를 보여줌
- [x] 조건이 충족되면 에러 메세지가 사라지도록 함

#### 로그인 - Password
- [x] PW 는 반드시 입력 되어야 하고 최소 8글자 이상 입력되어야 함
- [x] PW input field를 벗어났을 때, 이 조건이 충족되지 않으면 적절한 에러 메세지를 보여줌

#### 로그인 - 버튼
- [x] ID의 조건과 PW의 조건이 모두 충족되었을 때만 버튼이 활성화되도록 함

#### 리팩터링
- [x] 재사용을 고려해 커스텀 훅으로 로그인 기능을 구현할 것

---

멘토님께서 피드백 주신 것을 반영하여 버튼 컴포넌트를 수정해보았습니다. 일단 현재 구현한 버튼 3가지를 두고 생각해봤는데, 크게 < 기본 노란 버튼, 아웃 라인 버튼 / 크기 / 활성화 여부 > 로 나눌 수 있을 것 같아 이 기준을 잡고 수정했습니다. 아직까진 버튼이 많지 않아서 이정도로 커버가 되는 것 같은데, 버튼이 많아지면 어떤식으로 버튼 컴포넌트를 관리해야 할 지.. 예를 들면.. 어느 정도 비슷한 모양새를 하고 있지만 세세하게 차이점이 있는 버튼들의 경우 어떻게 처리를 하면 좋을지? 이 부분을 좀 더 고민해봐야 할 것 같아요..  🥲

daaoooy committed:
remove : 커밋 실수로 인해 생겨난 중복 파일 삭제

- 지난번 파일명 및 폴더구조 변경 후 커밋 실수로 인해서 생겨난 중복 파일을 삭제했습니다.

daaoooy committed:
feat(button) : ID, PW 조건 모두 만족해야 버튼이 활성화되는 기능 추가

- activeBackground, inactiveBackground (color 도 동일하게) 을
  이용하여 버튼 활성화 기능을 구현하였습니다.
- disable 값을 통해서 참이면 inactive, 거짓이면 active 로 되도록
  해주었습니다.
- 커서도 동일하게 disable을 통해 active일 때만 커서가 허용되도록
  해주었습니다.

daaoooy committed:
feat(hook) : 로그인 기능을 위한 hook 구현

- Email과 Password 조건들, 그리고  로그인 가능 여부를 체크합니다.
- useEffect를 통해 입력의 상태가 변경될 때마다 렌더링되도록
  해주었습니다.

daaoooy committed:
feat(login) : custom hook을 사용하여 로그인 기능 및 UI 구현

- 구현한 커스텀 훅 useLoginForm을 이용해 로그인 기능을 구현했습니다.
- 기존 UI에 에러 메세지가 나올 수 있도록 추가하였습니다.
- 에러 메세지는 에러가 있을 땐 보이도록, 에러가 없을 땐 보이지 않게(공간 차지X) 해주었습니다.
- 그 외에 세세한 UI 를 수정하였습니다 (input form 밑 보더 색상 변경,
  마진 조정 등)

daaoooy committed:
refactor : 컨벤션 통일을 위한 수정

yujo11 commented on src/component/Button/Button.tsx:
```diff
@@ -2,7 +2,7 @@ import type { ComponentPropsWithoutRef } from 'react'
 import styled from '@emotion/styled'
 
-type ButtonType = 'login' | 'More' | 'home'
+type ButtonType = 'login' | 'more' | 'home'
```
ButtonType을 보면 지금 Button Component가 Top-Down 방식으로 구현 된 것으로 보여요. (페이지를 먼저 만들고 버튼을 구현)

지금과 같은 구현방식이면 새로운 Button이 필요 할 때마다 매번 Button 컴포넌트의 type과 style이 추가되어야 할거 같아요. 정해진 type에 대응하는 대신에 Button 컴포넌트를 유용하게 사용할 수 있는 prop을 받는 방식으로 구현을 해보셔도 좋을거 같네요

디자인 시스템들에서는 어떤 형식으로 Button 컴포넌트를 제공하는지도 참고해보셔도 좋을거 같습니다!

- Antd: https://ant.design/components/button
- shardcn: https://ui.shadcn.com/docs/components/button

daaoooy replied:
주신 사이트를 참고해서 Button 컴포넌트를 variant, size, disabled props를 받는 방식으로 수정 완료했습니다.

yujo11 commented on src/features/Login/hooks/useLoginForm.ts:
```diff
@@ -0,0 +1,71 @@
+  }, [password, passwordTouched])
+
+  const loginOK = () => {
+    return emailRegEx.test(email) && password.length >= 8
```
매직넘버 대신 상수를 정의해서 사용해보셔도 좋을거 같아요

daaoooy replied:
PASSWORD_MIN_LENGTH 로 수정하였습니다.

yujo11 commented on src/features/Login/hooks/useLoginForm.ts:
```diff
@@ -0,0 +1,71 @@
+    /^[A-Za-z0-9]([-_.]?[A-Za-z0-9])*@[A-Za-z0-9]([-_.]?[A-Za-z0-9])*\.[A-Za-z]{2,3}$/i
+
+  const checkEmail = () => {
+    if (!emailTouched) return false
```
이 부분은 `checkEmail`과는 무관해보이는 로직인데요. 이 함수를 호출하는 시점에 체크해서 emailTouched가 true일 때만 이 함수를 호출해도 좋을거 같아요

daaoooy replied:
Touched 함수가 true 일 때만 check 함수를 호출하도록 수정하였습니다.

yujo11 reviewed (changes requested):
안녕하세요 다연님 ~ 소소한 코멘트들 남겼으니 확인해보시면 좋을거 같습니다

>작동은 잘 되는 것 같은데, 이런 식으로 코드를 짜는게 맞는지 항상 의문입니당.. 매번 너무 주먹구구식이 아닌가 싶어요. 이번 과제에서는 커스텀 훅과 그 훅 안에서 useState, useEffect 를 사용해서 구현해주었는데, 혹시 추가로 공부하고 고려해서 구현하면 좀 더 깔끔하고 이해하기 쉽게 구현할만한 react hook이 있을까요? 약간 "이런 상황에선 이러한 방법 (이러한 훅) 보다는 이런 것을 사용하면 좋아요!" 같은 것이 있을까요 .. ? !

hook을 구현하면서 고민했던 지점을 조금 더 구체적으로 설명해주시면 좋을거 같습니다. 지금은 어떤 부분을 고민하고 계신지 잘 모르겠네요 ㅎㅎ 

코드를 작성하는 것도 글쓰기와 비슷해서 많이 보고, 많이 작성해보고, 많이 생각해보는 것이 중요하다고 생각하는데요. 지금처럼 모두가 같은 요구사항을 가지고 구현을 진행하고 있는 상황은 코드를 보기 정말 좋은 상황이라고 생각해요. 같이 과정을 진행하고 있는 다른 사람들은 주어진 요구사항을 구현하기 위해 어떻게 코드를 작성했는지 PR을 살펴보시면 도움이 많이 될거라고 생각합니다 ㅎㅎ 

---

미션 진행하느라 고생 많으셨습니다 ~~ 👍 👍 👍

daaoooy committed:
refactor(button) : 버튼 컴포넌트 단점 개선

- 이전 구현 방식(Top-Down 형태, 로그인, 더보기 등과 같은 타입을 기준으로 버튼을 스타일하는 방식)에서
  variant, size, 활성화 여부(disabled), 버튼 크기 확장 여부(fullWidth) 라는 큰 기준에 따라 버튼을 스타일링하는 방식으로 변경하였습니다.
- 나누는 기준 (받아올 props) 에 가장 큰 변동사항이 있습니다.
  (이전 : [login, more, home] -> 현재 : [Variant: primary | outlined, Size: large | medium | small, disabled, fullWidth]
- setButtonStyle 을 추가하였습니다. ->  props를 받아 버튼 컴포넌트에 적용할 스타일값을 지정합니다.

daaoooy committed:
refactor(button) : Button 컴포넌트 변경에 따라 전달 props 재설정

- Button 컴포넌트 변경에 맞추어 전달할 prop를 변경해주었습니다.
- 추가적으로, NotFound의 subTitle 마진 값을 조정해주었습니다.

daaoooy committed:
refactor(hook) : 코드 피드백 반영 (무관한 로직 변경, 상수 치환)

- checkEmail 과 checkPassword 내의 무관한 로직 호출 위치를 수정하였습니다.
- 비밀번호 최소 길이인 8을 PASSWORD_MIN_LENGTH 로 만들어주었습니다.

yujo11 commented on src/component/Button/Button.tsx:
```diff
@@ -1,86 +1,138 @@
+type Size = 'large' | 'medium' | 'small'
 
 interface ButtonProps extends ComponentPropsWithoutRef<'button'> {
-  buttonType: ButtonType
+  variant?: Variant
+  size?: Size
+  disabled?: boolean
+  fullWidth?: boolean
 }
```
👍 👍 👍

yujo11 commented on src/features/Login/hooks/useLoginForm.ts:
```diff
@@ -1,4 +1,5 @@
 import { useState, useEffect } from 'react'
+const PASSWORD_MIN_LENGTH = 8
```
👍 👍 👍

yujo11 reviewed (approved):
안녕하세요 다연님 ~

피드백 반영해주신 내용 확인 했습니다! 고생 많으셨어요 다음 step에서 다시 뵈어요 ~~ 👍 👍 👍
